# 1. Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Set system path

In [2]:
# set path for data files
path = r'C:\Users\Stony\OneDrive\CareerFoundry\Data Immersion Course\Instacart Basket Analysis'

# 3. Import Products and Orders_wrangled into dataframes

In [3]:
# Import Products.csv into a dataframe
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import Orders_wrangled.csv into a dataframe
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 4. Data consistency checks

In [5]:
# Run descriptive statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

# 5. Create a mixed-type dataframe

In [6]:
# Create a dataframe called test
df_test = pd.DataFrame()

In [7]:
# Create a column in our test dateframe called mix and give it the following values: a, b, 1, True
df_test['mix'] = ['a', 'b', 1, True]

In [8]:
# Validate that the column and data was created successfully
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed-type columns in our test dataframe
# NOTE: updated version of Jupyter cannot use dataframe.applymap as function has been depeciated. Must use df.map instead
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Correct mix-type data by converting all to string
df_test['mix'] = df_test['mix'].astype('str')

# 6. Missing Values

In [11]:
# Sum the total number of blank or missing values in the products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create a new dataframe called nan with only the items from the products dataframe with missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
# Validate the dataframe was created and display the information on the items missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# Create a new dataframe called prods_clean with the items missing values removed
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# Validate the rows of missing data have been siccessfully removed from the dataframe
df_prods_clean.shape

(49677, 5)

# 7. Duplicate values

In [16]:
# Create a new dataframe called dups containing all duplicated records from the prods_clean dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
# Validate the dataframe was created and contains only items that are duplicated in the prods_clean data
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
# Validate the number of columns and rows in the prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [19]:
# Create a new dataframe called prods_clean_no_dups from prods_clean without any duplicated rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
# Validate the number of columns and rows in the prods_clean_no_dups dataframe to ensure the duplicated rows were removed
df_prods_clean_no_dups.shape

(49672, 5)

# 8. Export cleaned file

In [21]:
# Export cleaned products dataframe to a csv file with an easily understandable name
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))

# Exercise 4.5 Tasks

# 1. Consistency checks on Products dataframe

In [22]:
# Use the describe function to analyze the prods_clean_no_dups dataframe
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

There appears to be an issue with the max price in the dataframe, 9,999 is either an error in date entry or potentially a placeholder for the correct cost that is waiting to be or has not been updated yet. Also the product_id column has a count of 49,672 and a max value of 49,688 which can point to the presence of duplicate data in the dataframe.

# 2. Perform a descriptive analysis on the orders dataframe

In [23]:
# Use the describe function to analyze the orders dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

In the column days_since_last_order, there appear to be missing values as the count of that column is less than every other column. 
Also, the largest number in that the column (the max) is 30, which could mean we are only tracking up to 30 days of data or every customers always orders
at least once per 30 days. This would need to be answered by the customer to validate their data capture and tracking durations. In the order_number column, the max of 100 may mean that Instacart only captures the first 100 orders per customer or it is a rolling 100 orders and every order over 100 will lose earlier orders. This also needs to be answered by Instacart to confirm their data capture and data retention timeframes.

# 3. Check for mixed-type data in the orders dataframe

In [24]:
# Check for mixed-type columns in our orders dataframe
# Reminder: updated versions of Jupyter cannot use dataframe.applymap from the lesson as that function has been depeciated. Must use df.map instead
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There does not appear to be any mixed-type data in the orders dataframe

# 5. Check for missing values in the orders dataframe

In [25]:
# Sum the total number of blank or missing values in the orders dataframe
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206,209 missing values in the days_since_last_order column. This could be do to either an error in data collection with Instacart or it could be the number of first time users to the Instacart service as any new users will not have an order history and therefore have no prior orders.

# 6. Address missing values

In [26]:
# Create a new dataframe containing only the null values
df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [27]:
# Validate creation and contents of new dataframe
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [28]:
# Investigate if all of the order_numbers are 1 for all orders with a missing days_since_last_order
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [29]:
# Since all missing days since last order have an order number of 1, we must keep the data. Creating a column called first order will flag those rows
# and any subsequent new rows that are added after this analysis.
df_ords['first_order'] = df_ords['days_since_last_order'].isna()

In [30]:
# Validate the column first_orders was created successfully
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_last_order  first_order  
0                        8                    NaN         True  
1                        7                   15.0        False  
2                       12                   21.0        False  
3                        7                   29.0        False  
4                       15                   28.0        False  
...                    ...                    ...          ...  
3421078                 18                   29.0        False  
3421079                 10                   30.0        False  
3421080                 12                   18.0        False  
3421081                 12                    7.0        False  
3421082                 14                   30.0        False  

[3421083 rows x 8 columns]

In [31]:
# Validate the number of missing rows is the same after the creation of our new column
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
first_order                   0
dtype: int64

The missing values have been identified and addressed. It would not have been appropriate to drop the missing data because it is accurate and filtering or removing that data could have an effect on our analysis.

# 7. Check for duplicate values

In [32]:
# Create a new dataframe called dups containing all duplicated records from the orders dataframe
df_ord_dups = df_ords[df_ords.duplicated()]

In [33]:
df_ord_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_last_order, first_order]
Index: []

There do not appear to be any duplicated rows in the orders dataframe. If duplicates were present in the dataframe, I would use the folowing code 
to remove them: df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

# 9. Export cleaned dataframes into csv files

# Export cleaned products dataframe to a csv file with an easily understandable name
df_mprods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_cleaned.csv'))

# Export cleaned orders dataframe to a csv file with an easily understandable name
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [34]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0           0   2539329        1             1                  2   
1           1   2398795        1             2                  3   
2           2    473747        1             3                  3   
3           3   2254736        1             4                  4   
4           4    431534        1             5                  4   

   order_hour_of_day  days_since_last_order  first_order  
0                  8                    NaN         True  
1                  7                   15.0        False  
2                 12                   21.0        False  
3                  7                   29.0        False  
4                 15                   28.0        False

In [35]:
df_ords.drop(columns = ('Unnamed: 0'), inplace = True)

In [36]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'first_order'],
      dtype='object')

In [37]:
# Export cleaned orders dataframe to a csv file with an easily understandable name
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))